### PR TITLE
SWC-6734: Use legacy uploader for Google Cloud Platform bucket upload…

### PR DESCRIPTION
**Pull request by jay-hodgson**
…, and new SRC uploader for S3 (and S3-like) upload

### CODE REVIEW

**Comment by jay-hodgson** [comment]
Manually cherry-pick changes from https://github.com/Sage-Bionetworks/SynapseWebClient/pull/5333 so that the new code  can exist with the old implementation

**Comment by jay-hodgson** [comment]
The MultipartUploaderImpl in the previous PR is now MultipartUploaderImplV2.

File: src/main/java/org/sagebionetworks/web/client/widget/entity/download/Uploader.java [hunk 1]
```diff
@@ -669,14 +674,18 @@ public void directUploadStep2(String fileName) {
...
-          contentType,
-          currentFile,
-          this,
-          storageLocationId,
-          view
-        );
+      // TODO: PLFM-8252: If Google Cloud platform Synapse solution supports parallel upload, then remove legacyMultiPartUploader (and all associated code)
+      MultipartUploader currentUploader = currentUploadType == UploadType.S3
+        ? multiPartUploader
+        : legacyMultiPartUploader;
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/download/Uploader.java:680]: If S3 then use V2, else use V1

File: src/test/java/org/sagebionetworks/web/unitclient/widget/upload/MultipartUploaderV2Test.java [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+    ).thenReturn(mockPromise);
+
+    when(synapseJsniUtils.getFileSize(any(JavaScriptObject.class))).thenReturn(
+      FILE_SIZE
+    );
+    // fire the timer
+    CallbackMockStubber.invokeCallback()
+      .when(gwt)
+      .scheduleExecution(any(Callback.class), anyInt());
+    when(gwt.createXMLHttpRequest()).thenReturn(null);
```
Comment (nickgros) [inline src/test/java/org/sagebionetworks/web/unitclient/widget/upload/MultipartUploaderV2Test.java]: Can some of these mocks be removed?

File: src/test/java/org/sagebionetworks/web/unitclient/widget/entity/download/UploaderTest.java [hunk 3]
```diff
@@ -238,28 +236,27 @@ public void before() throws Exception {
...
+      mockSynapseJsniUtils,
+      mockGwt,
+      mockAuthenticationController,
+      multipartUploader,
+      mockGlobalApplicationState,
+      mockS3DirectUploader,
+      mockSynapseJavascriptClient,
+      mockSynapseProperties,
+      mockEventBus,
+      mockGinInjector
```
Comment (nickgros) [inline src/test/java/org/sagebionetworks/web/unitclient/widget/entity/download/UploaderTest.java:264]: Can you add a test to verify that the legacy uploader is used when using Google? Sorry if I missed it, this code reformatting is making things much harder

File: src/test/java/org/sagebionetworks/web/unitclient/widget/upload/MultipartUploaderV2Test.java (hunk 2 above)
Comment (jay-hodgson) [reply src/test/java/org/sagebionetworks/web/unitclient/widget/upload/MultipartUploaderV2Test.java]: Done

File: src/test/java/org/sagebionetworks/web/unitclient/widget/entity/download/UploaderTest.java (hunk 3 above)
Comment (jay-hodgson) [reply src/test/java/org/sagebionetworks/web/unitclient/widget/entity/download/UploaderTest.java:264]: Done

**Review by nickgros — APPROVED** [review]

**Comment by jay-hodgson** [comment]
Depends on SRC 3.2.7

**Pull request merged**